### ISSUE TITLE
Not having expected performance on distance matrix module

### ISSUE THREAD

**Issue by srimaln91** [valhalla/valhalla#2604]
We are evaluating Valhalla to replace OSRM since it provides dynamic routing capabilities. So I have conducted a performance test to compare the performance of road matrix module in OSRM and Valhalla. Below are the results.

Server Specs: 4 vCPUs 2.8GHz, 8GB memory and 20GB SSD.
OSM FIle: [http://download.geofabrik.de/asia/sri-lanka-latest.osm.pbf](http://download.geofabrik.de/asia/sri-lanka-latest.osm.pbf)

**Routing**
Engine | Throughput/s | Average | 90th | 95th | 99th | Max
-- | -- | -- | -- | -- | -- | --
OSRM | 700 | 130 | 36 | 53 | 142 | 183
Valhalla | 1000 | 13 | 21 | 53 | 99 | 194

As per the above results, the Valhalla routing submodule is very performant than OSRM. But we were unable to get expected performance on distance matrix module.

We were running distance matrix test with 33 sources and one destination on **sources_to_targets** endpoint. 

All duration values are in milliseconds

**DIstance Matrix**
Engine | Throughput/s | Average | 90th | 95th | 99th | Max
-- | -- | -- | -- | -- | -- | --
OSRM | 100 | 10 | 12 | 13 | 17 | 44
Valhalla | 45 | 2095 | 3992 | 4300 | 4624 | 4986

I would like to know if you have any suggestions to improve the performance of distance matrix calculations.

Thank You.

**Comment by danpat** [comment]
As someone who has contributed to both projects - there is something very wrong with your routing results for OSRM - 130ms average response time is crazy slow.  You should debug your system to see if you have a bottleneck somewhere (disk, network, CPU, idk).  OSRM should be at *least* on a par with Valhalla for single route requests, and should greatly outperform it for long-distance routes.

The obvious performance difference in the matrix interfaces highlights this - you should be seeing a similar, but not quite so extreme, ratio of response times.  Your OSRM matrix numbers are more in line with what I expect.

In *general*, OSRM is faster, but the price paid is flexibility.  If you need very fast results, use OSRM.  If you need to supply variable routing constraints/costings per-query, use Valhalla.  Unfortunately, you [can't have your cake and eat it too](https://en.wikipedia.org/wiki/You_can%27t_have_your_cake_and_eat_it).

**Comment by srimaln91** [comment]
Thanks, @danpat for the quick response.

Let me check if anything goes wrong in my OSRM setup.

**Comment by srimaln91** [comment]
Hi Daniel,

I've executed the load test again with some changes and results are below.

**Routing**
Engine | Throughput/s | Average Response Time | 90th | 95th | 99th | Max
-- | -- | -- | -- | -- | -- | --
OSRM | 510 | 2 | 3 | 3 | 15 | 27
Valhalla | 220 | 379 | 564 | 586 | 614 | 632

We were running distance matrix test with 33 sources and one destination on **sources_to_targets** endpoint. 

**Distance Matrix**
Engine | Throughput/s | Average Response Time | 90th | 95th | 99th | Max
-- | -- | -- | -- | -- | -- | --
OSRM | 100 | 23 | 30 | 36 | 55 | 106
Valhalla | 15 | 5139 | 7033 | 7205 | 7379 | 7383

As you said, OSRM is quite performant on both modules. We've decided to run multiple OSRM instances for different routing profiles. Also, we are looking forward to using Valhalla on some specific use-cases which require dynamic routing capabilities.

I would like to know if you have any suggestions for improving the performance of Valhalla.

Thank you again for the valuable input.

**Comment by Doidel** [comment]
I would also love to use Valhalla for distance matrix calculations, but we have up to 10'000 x 10'000 destinations. From this small benchmark this seems inapproachable for Valhalla. Is this correct, and is this still the case?

**Comment by kevinkreiser** [comment]
It all depends on how close the locations are to each other. The further apart any one location is from the cluster the more un approachable it becomes

**Comment by Doidel** [comment]
I have just checked: Furthest distance is 250km*. That's the upper bound.
Should we even bother with making our own benchmark? (We'll need the result of 10k x 10k in minutes, not hours).
Thanks for your advice!

*(46.10046, 10.940258), (47.591244, 8.47405)

**Comment by kevinkreiser** [comment]
i would advise you to always test. its hard for me to guess your hardware and the performance of the algorithm. i suspect it will take a while to compute that but at the same time you can speed it up by breakign the request into many requests and doing them in parallel if you have a machine with multiple cores. i dont want you to waste your time but i cannot give you a definitive answer on whether it will meet your requirements or not

**Comment by MichaelKorn** [comment]
I compared some OSM based routing engines. I also compared the response times for OSRM (very fast / not flexible / missing features) and Valhalla (slow / flexible / more features).
- Experiments from Mai 2022
- I used (loaded) the map of Germany
- I used the public docker images
- Notebook with AMD Ryzen 7 4800H
- Error bars show 3sigma (there are also error bars for OSRM, but covered by the dots)

Random locations in Bonn (this means shorter distances):
![Screenshot 2022-11-08 224742](https://user-images.githubusercontent.com/3801648/200683387-3aa38dd8-b674-427d-90dc-3d147a31a473.png)

Random locations in Germany (As already pointed out I needed to overwrite the Valhalla default "Furthest distance" value)
![Screenshot 2022-11-08 224854](https://user-images.githubusercontent.com/3801648/200683810-8a6d1982-caeb-4b61-a0ad-1d8952b02711.png)
I also measured 1k, 2k, 2.5k random location within Germany for OSRM: 12s, 33s, 46s. In the meantime, the good response times have also been confirmed in regular use.
I did not systematic measurements of larger matrices for Valhalla because it took too long. 175² already leads to response times > one minute, e.g. this is the default timeout of nginx and was no acceptable for the application.

**Comment by DavidPiringer** [comment]
I just want to share my current experience with big distance matrices and Valhalla.

Calculation of two matrices (3855 x 3855 and 12425 x 12425) for Switzerland (bounding box: `47.21400 7.99403; 47.62118 8.72132`) with the following results:
- It's not possible to calculate the whole matrix due to too high RAM usage (>50gb).
- The RAM problem is solvable with batches (e.g. 200 x 200 batches and 1 x 500), but a lot of requests are necessary. My results regarding the calculation time were:
  - 3855 x 3855: about 10 Minutes
  - 12425 x 12425: calculation manually cancelled after 4.5 hours

We ran Valhalla in an docker container, based on the gisops image (3.2.0) with the following WSL configuration:
- 19gb RAM
- 200gb SWAP

Hardware:
- Lenovo ThinkPad P15v
- Intel Core i7-12700H
- 32gb DDR5 RAM

Sadly, Valhalla is currently not usable for our usecase.

**Comment by nilsnolde** [comment]
Thanks for the stats!

> 3855 x 3855: about 10 Minutes

Not that bad at all! Honestly I'd have assumed worse results than that. But surely not a good timing if you need instant feedback. Was it locations distributed over entire Switzerland or only located in a subset of Switzerland?

> It's not possible to calculate the whole matrix due to too high RAM usage

Yeah that's the many:many algo which expands from all locations at the same time, so you basically load the whole Switzerland network into RAM for a single request. While the one:many (timedistancematrix) algo will not do that, it'll also be a pretty much a magnitude slower than many:many (costmatrix).

> The RAM problem is solvable with batches (e.g. 200 x 200 batches and 1 x 500)

Not sure if you tried, but this can also be done in parallel. RAM usage will be far less than just one request, but still considerable. Smaller batches are best done without HTTP, either by doing a C++ script or Python using the bindings.

**Comment by DavidPiringer** [comment]
Thanks for the reply @nilsnolde!

> Not that bad at all! Honestly I'd have assumed worse results than that. But surely not a good timing if you need instant feedback. Was it locations distributed over entire Switzerland or only located in a subset of Switzerland?

The locations are located in the cantons of Lucerne, Aargau, Solothurn and Zurich.

> Not sure if you tried, but this can also be done in parallel. RAM usage will be far less than just one request, but still considerable. Smaller batches are best done without HTTP, either by doing a C++ script or Python using the bindings.

I forgot to mention, the requests were done in parallel. The WSL machine has 10 cores, and we made sure that always 10 requests were calculating in parallel. Unfortunately, the C++ or Python implementations will not create the speedup we need.

**Comment by nilsnolde** [comment]
If you're _very_ serious about making this work, it should be possible to return such an instance in a reasonable amount of time, but definitely comes at the price of quite a big infrastructure of multiple Valhalla servers with as many threads as possible. Just leaving it here, I'd imagine for _someone_ it might actually be worth the expense, to have highly dynamic matrix results.

**Comment by kevinkreiser** [comment]
@ktnr yep that's exactly what @nilsnolde is referring to. Valhalla does only a very very conservative contraction on the graph so that we get a doubling of speed over basic A* algorithms but can still do a full edge weight customization. With osrm you take contraction to the max essentially fixing the edge weights but making extracting a path very efficient and exact for a single configuration. Other than this fundamental difference the only optimizations are more general like caching, parallelism, avoiding useless computations etc basically doing work to profile the code and making changes where we see the most cpu being spent
